src/graphinfo.jl

```diff
@@ -0,0 +1,254 @@
+using AbstractPPL
+import Base.getindex
+using SparseArrays
+using Setfield
+using Setfield: PropertyLens, get
+
+"""
+    GraphInfo
+
+Record the state of the model as a struct of NamedTuples, all
+sharing the same key values, namely, those of the model parameters.
+`value` should store the initial/current value of the parameters.
+`input` stores a tuple of inputs for a given node. `eval` are the
+anonymous functions associated with each node. These might typically
+be either deterministic values or some distribution, but could an 
+arbitrary julia program. `kind` is a tuple of symbols indicating
+whether the node is a logical or stochastic node. Additionally, the 
+adjacency matrix and topologically ordered vertex list and stored.
+
+GraphInfo is instantiated using the `Model` constctor. 
+"""
+
+struct GraphInfo{T} <: AbstractModelTrace
+    value::NamedTuple{T}
+    input::NamedTuple{T}
+    eval::NamedTuple{T}
+    kind::NamedTuple{T}
+    A::SparseMatrixCSC
+    sorted_vertices::Vector{Symbol}
+end
+
+"""
+    Model(;kwargs...)
+
+`Model` type constructor that takes in named arguments for 
+nodes and returns a `Model`. Nodes are pairs of variable names
+and tuples containing default value, an eval function 
+and node type. The inputs of each node are inferred from 
+their anonymous functions. The returned object has a type 
+GraphInfo{(sorted_vertices...)}.
+
+# Examples
+```jl-doctest
+julia> using AbstractPPL
+
+julia> Model(
+               s2 = (0.0, () -> InverseGamma(2.0,3.0), :Stochastic), 
+               μ = (1.0, () -> 1.0, :Logical), 
+               y = (0.0, (μ, s2) -> MvNormal(μ, sqrt(s2)), :Stochastic)
+           )
+Nodes: 
+μ = (value = 1.0, input = (), eval = var"#6#9"(), kind = :Logical)
+s2 = (value = 0.0, input = (), eval = var"#5#8"(), kind = :Stochastic)
+y = (value = 0.0, input = (:μ, :s2), eval = var"#7#10"(), kind = :Stochastic)
+```
+"""
+
+struct Model{T}
+    g::GraphInfo{T}
+end
+
+function Model(;kwargs...)
+    vals = getvals(NamedTuple(kwargs))
+    args = [argnames(f) for f in vals[2]]
+    A, sorted_vertices = DAG(NamedTuple{keys(kwargs)}(args))    
+    modelinputs = NamedTuple{Tuple(sorted_vertices)}.([vals[1], Tuple.(args), vals[2], vals[3]])
+    Model(GraphInfo(modelinputs..., A, sorted_vertices))
+end
+
+
+"""
+    DAG(inputs)
+
+Function taking in a NamedTuple containing the inputs to each node 
+and returns the implied adjacency matrix and topologically ordered 
+vertex list.
+"""
+function DAG(inputs)
```

Comment:
It’s unusual to use captical function names, perhaps
```suggestion
function dag(inputs)
```

Reply:
Just changed this